File: CodeMatch/Models/HistoryModels.swift
import Foundation

struct MatchHistoryEntry: Identifiable, Codable, Equatable {
    let id: UUID
    let code: String
    let matchedAt: Date
    /// 照合時のQRコード全文。旧バージョンで記録した履歴ではnil。
    var qrPayload: String?
    /// 照合時のCode 128全文。旧バージョンで記録した履歴ではnil。
    var barcodePayload: String?

    init(
        id: UUID = UUID(),
        code: String,
        matchedAt: Date = Date(),
        qrPayload: String? = nil,
        barcodePayload: String? = nil
    ) {
        self.id = id
        self.code = code
        self.matchedAt = matchedAt
        self.qrPayload = qrPayload
        self.barcodePayload = barcodePayload
    }
}

struct MatchSession: Identifiable, Codable, Equatable {
    let id: UUID
    let startedAt: Date
    var endedAt: Date?
    var entries: [MatchHistoryEntry]
    /// 任意のセッション名。旧バージョンの履歴や未入力ではnil。
    var name: String?

    init(
        id: UUID = UUID(),
        startedAt: Date = Date(),
        endedAt: Date? = nil,
        entries: [MatchHistoryEntry] = [],
        name: String? = nil
    ) {
        self.id = id
        self.startedAt = startedAt
        self.endedAt = endedAt
        self.entries = entries
        self.name = name
    }

    var isActive: Bool { endedAt == nil }
    var matchedCount: Int { entries.count }

    /// 表示用のセッション名。未設定なら空文字。
    var displayName: String { name?.trimmingCharacters(in: .whitespacesAndNewlines) ?? "" }

    /// この品番がこのセッションで照合された回数。1回の照合を1箱として数える。
    func matchCount(code: String) -> Int {
        entries.filter { $0.code == code }.count
    }

    /// 同一品番の照合をまとめたグループ。最初に照合された順に並ぶ。
    var groupedEntries: [GroupedMatchEntry] {
        var order: [String] = []
        var buckets: [String: [MatchHistoryEntry]] = [:]
        for entry in entries {
            if buckets[entry.code] == nil { order.append(entry.code) }
            buckets[entry.code, default: []].append(entry)
        }
        return order.map { GroupedMatchEntry(code: $0, entries: buckets[$0] ?? []) }
    }
}

/// 同一品番の照合履歴を1つにまとめた表示用グループ。1件の照合 = 1箱。
struct GroupedMatchEntry: Identifiable, Equatable {
    let code: String
    /// 照合順（古い順）の個別記録。
    let entries: [MatchHistoryEntry]

    var id: String { code }
    var boxCount: Int { entries.count }
    var firstMatchedAt: Date { entries.first?.matchedAt ?? .distantPast }
    var lastMatchedAt: Date { entries.last?.matchedAt ?? .distantPast }
}

/// 履歴表示用の曜日付き日時フォーマッタ (例: 2026/08/17(日) 21:35)
enum JPDate {
    private static let dateTimeFormatter: DateFormatter = {
        let formatter = DateFormatter()
        formatter.locale = Locale(identifier: "ja_JP")
        formatter.dateFormat = "yyyy/MM/dd(E) HH:mm"
        return formatter
    }()

    private static let timeFormatter: DateFormatter = {
        let formatter = DateFormatter()
        formatter.locale = Locale(identifier: "ja_JP")
        formatter.dateFormat = "HH:mm"
        return formatter
    }()

    static func dateTime(_ date: Date) -> String {
        dateTimeFormatter.string(from: date)
    }

    static func time(_ date: Date) -> String {
        timeFormatter.string(from: date)
    }
}

File: android/core/model/src/main/kotlin/jp/rimtty/codematch/core/model/HistoryModels.kt
package jp.rimtty.codematch.core.model

import java.util.UUID

/**
 * One successful comparison recorded as one inspected box.
 *
 * IDs are UUID strings and timestamps are UTC epoch milliseconds so the model
 * can be persisted by Room without coupling it to Android date/time classes.
 */
data class MatchEntry(
    val id: String = UUID.randomUUID().toString(),
    val code: String,
    val matchedAt: Long = System.currentTimeMillis(),
    val qrPayload: String? = null,
    val barcodePayload: String? = null,
    /** Monotonic position in the parent session, starting at zero. */
    val sequence: Long = 0L,
)

/** A comparison session, including only successful match entries. */
data class MatchSession(
    val id: String = UUID.randomUUID().toString(),
    val startedAt: Long = System.currentTimeMillis(),
    val endedAt: Long? = null,
    val entries: List<MatchEntry> = emptyList(),
    val name: String? = null,
) {
    val isActive: Boolean
        get() = endedAt == null

    val matchedCount: Int
        get() = entries.size

    /** A blank or whitespace-only name is displayed as no name. */
    val displayName: String
        get() = name?.trim().orEmpty()

    /** Number of boxes whose recorded part number equals [code]. */
    fun matchCount(code: String): Int = entries.count { it.code == code }

    /**
     * Groups duplicate part numbers in first-seen order. Each entry remains in
     * scan order so the group represents individual boxes rather than a set of
     * unique part numbers.
     */
    val groupedEntries: List<GroupedMatchEntry>
        get() {
            val buckets = LinkedHashMap<String, MutableList<MatchEntry>>()
            entries.forEach { entry ->
                buckets.getOrPut(entry.code) { mutableListOf() }.add(entry)
            }
            return buckets.map { (code, grouped) ->
                GroupedMatchEntry(code = code, entries = grouped.toList())
            }
        }
}

/**
 * Result of attempting to finish a session.
 *
 * Empty sessions are intentionally removed from history, so callers can show
 * a different confirmation for that case than for a persisted ended session.
 */
sealed interface EndSessionOutcome {
    /** The session had no boxes and was discarded. */
    data class DeletedEmpty(val sessionId: String) : EndSessionOutcome

    /** The session had at least one box and now has an end timestamp. */
    data class Ended(val sessionId: String, val endedAt: Long) : EndSessionOutcome

    /** No session with the requested id (or no active session) exists. */
    data object NotFound : EndSessionOutcome

    /** The requested session had already been ended. */
    data class AlreadyEnded(val sessionId: String, val endedAt: Long) : EndSessionOutcome
}

/** A first-seen-order group of boxes carrying the same part number. */
data class GroupedMatchEntry(
    val code: String,
    val entries: List<MatchEntry>,
) {
    val id: String
        get() = code

    val boxCount: Int
        get() = entries.size

    val firstMatchedAt: Long
        get() = entries.firstOrNull()?.matchedAt ?: Long.MIN_VALUE

    val lastMatchedAt: Long
        get() = entries.lastOrNull()?.matchedAt ?: Long.MIN_VALUE
}
